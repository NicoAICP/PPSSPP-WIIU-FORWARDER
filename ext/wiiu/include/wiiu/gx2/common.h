#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
