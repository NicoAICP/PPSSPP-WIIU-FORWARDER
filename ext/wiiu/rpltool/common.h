#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
