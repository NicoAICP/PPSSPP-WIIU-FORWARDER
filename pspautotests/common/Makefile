TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
