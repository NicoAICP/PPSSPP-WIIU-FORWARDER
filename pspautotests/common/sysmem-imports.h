extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();