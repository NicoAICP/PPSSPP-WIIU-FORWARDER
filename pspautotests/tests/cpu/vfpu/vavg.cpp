#include <common.h>

#include "vfpu_common.h"

const ALIGN16 ScePspFVector4 vZero = {0, 0, 0, 0};
const ALIGN16 ScePspFVector4 vMinusOne = {-1, -1, -1, -1};

ALIGN16 ScePspFVector4 input, output;

#define GEN_SV(FuncName, Op, PFX) \
	void NOINLINE FuncName(ScePspFVector4 *v0, const ScePspFVector4 *v1) { \
	asm volatile ( \
	"lv.q   C100, %1\n" \
	Op " S000, "PFX"100\n" \
	"sv.q   C000, %0\n" \
	: "+m" (*v0) : "m" (*v1) \
	); \
}

#define GEN_PTQ(genFun, name) \
	genFun(name ## _p, #name ".p", "C"); \
	genFun(name ## _t, #name ".t", "C"); \
	genFun(name ## _q, #name ".q", "C");

GEN_PTQ(GEN_SV, vavg);
GEN_PTQ(GEN_SV, vfad);

void testSV(const char *desc, void (*vvvxxx)(ScePspFVector4 *v0, const ScePspFVector4 *v1)) {
	(*vvvxxx)(&output, &input);
	printVector(desc, &output);
}

void test_vavg_vfad(const char *subdesc) {
	char temp[256];

	sprintf(temp, "vavg.p: %s", subdesc);
	testSV(temp, &vavg_p);
	sprintf(temp, "vavg.t: %s", subdesc);
	testSV(temp, &vavg_t);
	sprintf(temp, "vavg.q: %s", subdesc);
	testSV(temp, &vavg_q);

	sprintf(temp, "vfad.p: %s", subdesc);
	testSV(temp, &vfad_p);
	sprintf(temp, "vfad.t: %s", subdesc);
	testSV(temp, &vfad_t);
	sprintf(temp, "vfad.q: %s", subdesc);
	testSV(temp, &vfad_q);
}

extern "C" int main(int argc, char *argv[]) {
	input.x = -0.0f;
	input.y = -0.0f;
	input.z = -0.0f;
	input.w = -0.0f;
	test_vavg_vfad("Negative zero");

	input.x = 1.0f;
	input.y = NAN;
	input.z = 1.0f;
	input.w = 1.0f;
	test_vavg_vfad("Single NAN");

	input.x = 1.0f;
	input.y = -INFINITY;
	input.z = 1.0f;
	input.w = 1.0f;
	test_vavg_vfad("Single -INFINITY");

	input.x = -1.0f;
	input.y = 1.0f;
	input.z = -1.0f;
	input.w = 1.0f;
	test_vavg_vfad("Sign near zero");

	return 0;
}