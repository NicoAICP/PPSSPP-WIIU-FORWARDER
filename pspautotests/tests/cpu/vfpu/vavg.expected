vavg.p: Negative zero: 0.000000,nan,nan,nan
vavg.t: Negative zero: 0.000000,nan,nan,nan
vavg.q: Negative zero: 0.000000,nan,nan,nan
vfad.p: Negative zero: 0.000000,nan,nan,nan
vfad.t: Negative zero: 0.000000,nan,nan,nan
vfad.q: Negative zero: 0.000000,nan,nan,nan
vavg.p: Single NAN: nan,nan,nan,nan
vavg.t: Single NAN: nan,nan,nan,nan
vavg.q: Single NAN: nan,nan,nan,nan
vfad.p: Single NAN: nan,nan,nan,nan
vfad.t: Single NAN: nan,nan,nan,nan
vfad.q: Single NAN: nan,nan,nan,nan
vavg.p: Single -INFINITY: -inf,nan,nan,nan
vavg.t: Single -INFINITY: -inf,nan,nan,nan
vavg.q: Single -INFINITY: -inf,nan,nan,nan
vfad.p: Single -INFINITY: -inf,nan,nan,nan
vfad.t: Single -INFINITY: -inf,nan,nan,nan
vfad.q: Single -INFINITY: -inf,nan,nan,nan
vavg.p: Sign near zero: 0.000000,nan,nan,nan
vavg.t: Sign near zero: -0.333333,nan,nan,nan
vavg.q: Sign near zero: 0.000000,nan,nan,nan
vfad.p: Sign near zero: 0.000000,nan,nan,nan
vfad.t: Sign near zero: -1.000000,nan,nan,nan
vfad.q: Sign near zero: 0.000000,nan,nan,nan
